### PR TITLE
hww.c: Turn OP_UNLOCK into an asynchronous operation.

### PR DESCRIPTION
**Pull request by conte91**
Make the OP_UNLOCK hww opcode start the unlock workflow asynchonously.
Support the retry/cancel flow during unlock.

### CODE REVIEW

File: src/hww.c [hunk 1]
```diff
@@ -149,28 +168,62 @@ static size_t _api_info(uint8_t* buf)
...
         case OP_UNLOCK:
             if (!memory_is_initialized()) {
-                out_rsp->data[0] = OP_STATUS_FAILURE_UNINITIALIZED;
+                out_rsp->buffer.data[0] = OP_STATUS_FAILURE_UNINITIALIZED;
+                out_rsp->buffer.len = 1;
+                return;
             } else {
-                out_rsp->data[0] =
-                    workflow_unlock_blocking() ? OP_STATUS_SUCCESS : OP_STATUS_FAILURE;
+                _state.unlock_wf = workflow_unlock(&_unlock_finished_cb, NULL);
```
Comment (benma) [inline src/hww.c]: workflow_unlock finishes immediately if the device is already unlocked. With the current retry logic, the unlock request is delayed by whatever retry-interval the client has, which is noticeable.

How to solve this?

File: src/hww.c [hunk 2]
```diff
@@ -149,28 +168,62 @@ static size_t _api_info(uint8_t* buf)
...
+ */
+static void _unlock(void)
+{
+    _clear_state();
+    if (_state.blocking_op != HWW_BLOCKING_OP_NONE) {
+        usb_processing_unlock();
+    }
+}
+
+static void _unlock_finished_cb(bool result, void* param)
```
Comment (benma) [inline src/hww.c]: is this ever called with `result=false` on cancel? not sure i understand how HWW_REQ_CANCEL leads to the proper (negative) response.

Comment (conte91) [reply src/hww.c]: The `HWW_REQ_CANCEL` is handled by the separate arbitration logic, which workflows know nothing about. The only case the unlock can fail (`result == false`) is if `memory_is_initialized() == false` when the unlock workflow is started.

`HWW_REQ_CANCEL` is handled internally by `hww.c` (see `_process_msg`).

File: src/hww.c (hunk 1 above)
Comment (conte91) [reply src/hww.c]: I would just add a check in `hww.c` that avoids calling `workflow_unlock` entirely if the device is already unlocked. The only other case in which the workflow will return immediately is the path with `!memory_is_initialized()`, which is rare enough (IMHO) to make a 200ms irrelevant. (Also, if `OP_UNLOCK` is only sent during pairing, I don't think 200ms would make a difference there as well, but that's again IMHO).

File: src/hww.c (hunk 2 above)
Comment (benma) [reply src/hww.c]: you mean `_cancel_packet`. thx, i see now

File: src/hww.c (hunk 1 above)
Comment (conte91) [reply src/hww.c]: Fixed in e9a5715

**Comment by benma** [comment]
Thanks for all the amazing async work and teaching. This PR is superseded by https://github.com/digitalbitbox/bitbox02-firmware/pull/442/, which implements almost the same, but in Rust.

**Pull request closed without merging**